### PR TITLE
[Contracts 1.2] Create a Kolibri Savings Pool

### PR DESCRIPTION
**Pull request by keefertaylor**
# Introduction

This PR is a rollup of features that implement the Kolibri Savings Rate, as defined by [KIP-010](https://discuss.kolibri.finance/t/kip-010-the-kolibri-savings-rate/79/2).

## Implementation

The Kolibri savings pool is a contract where users can deposit `kUSD` and receive `KSR`, a tracking token for their contributions. 

The pool pays a set amount of interest (requested from the stability fund) and thus over time the amount of `kUSD` in the pool grows. 

##  Functionality 

Liquidity Providers ("LPs") stake `kUSD` in the contract by calling `deposit`. In return they receive `KSR` tokens, which entitle to them to their share of the `kUSD` in the Savings Pool. Every time the savings pool gets interacted with, it calculates the amount of interest it is owed, and requests the interest from the Stability Fund. 

LPs can redeem `KSR` tokens by calling `redeem` at any time. The `KSR` tokens are burnt and the LP receives their share of `kUSD` back from the pool. 

## Pause Guardian

A special multisig, called the `pause guardian` can pause interest accruals in the fund in case of emergencies or an exploit. This functionality can only pause the system, it cannot access, control or steal user's funds. Only governance can restart the system. 

## Governance

A special contract, called the `governor` can make changes to the contract. The governor is intended to be the [Kolibri DAO](https://governance.kolibri.finance).  Specifically, the following parameters are governable in the savings pool contract:
- governorAddress: The governor may rotate itself to a new address (ex. a new multisig or a DAO)
- stabilityFundAddress: The address of the stability fund, where funds are requested from. 
- interestRate: The amount of interest accrued per deposit

On the Stability Fund, a new governable parameter is added:
- savingsPoolAddress: The address of a pool that is trusted to request funds. 

As stated above, the governor can unpause a system when the pause guardian has paused it.

Additionally, a number of mundane plumbing parameters are also updatable in order to support upgrades in Kolibri and Tezos metadata standards.
- contractMetadata: Updatable to allow compliance with with emerging standards
- tokenMetada: Updateable to allow compliance with emerging standards

## Interface

```
# Deposit a number of tokens and receive LP tokens back.
deposit(nat numTokens)

# Redeem a number of LP tokens for underlying tokens.
redeem(nat numLpTokens)
```

## Pull Requests

The individual PRs are:
#19 - Update Stability Fund to Interface with KSR
#21 - Implement KSR
#22 - Standardize Errors
#23 - Standardize Entrypoint Names 
#24 - Fix Schema for Token Metadata
#29 - Remove Manually Accrue Interest Entrypoint
#27 - Add pause guardian functionality
#31 - Upgrade SmartPy
#34 - Add off-chain views
#26 - Add migration scripts
#38 - Revert view code (Reverts #34)
#37 - Add Methods to zero out developer and stability funds
#28 - Fix contract metadata in deploys
#38 - Sanity check balances in deploys
#39 - Add a multisig to the deploy process
#44 - Infrastructure fixes from @Fitblip 
#45 - Final configuration updates
#46 - Fix verification step for funds

### CODE REVIEW

File: smart_contracts/savings-pool.py [hunk 1]
```diff
@@ -0,0 +1,4564 @@
...
+    savedState_tokensToDeposit = sp.none,
+    savedState_depositor = sp.none,
+  ):
+    token_id = sp.nat(0)
+
+    token_entry = sp.map(
+      l = {
+        "name": sp.bytes('0x496e7465726573742042656172696e67206b555344'), # Interest Bearing kUSD
+        "decimals": sp.bytes('0x3138'), # 18
+        "symbol": sp.bytes('0x69626b555344'), # ibkUSD
```
Comment (Fitblip) [inline smart_contracts/savings-pool.py:64]: Can we update this to `KSR` token? I think `ibkUSD` might be confusing 😬

File: smart_contracts/savings-pool.py [hunk 2]
```diff
@@ -0,0 +1,4564 @@
...
+        "icon": sp.bytes('0x2068747470733a2f2f6b6f6c696272692d646174612e73332e616d617a6f6e6177732e636f6d2f6c6f676f2e706e67') # https://kolibri-data.s3.amazonaws.com/logo.png
```
Comment (Fitblip) [inline smart_contracts/savings-pool.py:65]: I can generate you/host you a new icon/logo for this token 👍

File: smart_contracts/savings-pool.py [hunk 3]
```diff
@@ -0,0 +1,4564 @@
...
+    )
+    token_metadata = sp.big_map(
+      l = {
+        token_id: sp.record(token_id = 0, token_info = token_entry)
+      },
+      tkey = sp.TNat,
+      tvalue = sp.TRecord(token_id = sp.TNat, token_info = sp.TMap(sp.TString, sp.TBytes))
+    )
+
+    metadata_data = sp.bytes_of_string('{ "name": "Interest Bearing kUSD",  "description": "Interest Bearing kUSD",  "authors": ["Hover Labs <hello@hover.engineering>"],  "homepage":  "https://kolibri.finance", "interfaces": [ "TZIP-007-2021-01-29"] }')
```
Comment (Fitblip) [inline smart_contracts/savings-pool.py:78]: Same here, please update to KSR

File: smart_contracts/savings-pool.py [hunk 4]
```diff
@@ -0,0 +1,4564 @@
...
+    savedState_redeemer = sp.none,
+    savedState_tokensToDeposit = sp.none,
+    savedState_depositor = sp.none,
+  ):
+    token_id = sp.nat(0)
+
+    token_entry = sp.map(
+      l = {
+        "name": sp.bytes('0x496e7465726573742042656172696e67206b555344'), # Interest Bearing kUSD
+        "decimals": sp.bytes('0x3138'), # 18
```
Comment (Fitblip) [inline smart_contracts/savings-pool.py:63]: Also is this actually an 18 digit mantissa? I just got back something with 36 decimals 🤔

**Comment by keefertaylor** [comment]
Merging this now that Kolibri DAO proposal #10 has passed: https://governance.kolibri.finance/proposals/10

**Pull request merged**